openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -171,19 +169,66 @@ impl<'a, T: Clone + Debug> AbDiff<'a, T> {
         Ok(NodeId::try_from_node_index(self, node_index)?)
     }
 
-    /// Returns references to the leaves of the diff in order from left to
-    /// right. This function should not throw an error. However, it might throw
-    /// a [`LibraryError`] error if there is a bug in the implementation.
-    pub(crate) fn leaves(&self) -> Result<Vec<NodeId>, LibraryError> {
-        let mut leaf_references = Vec::new();
-        for leaf_index in 0..self.leaf_count() {
-            let node_index = to_node_index(leaf_index);
-            // The node reference must be valid, since it is a valid leaf
-            let node_ref = NodeId::try_from_node_index(self, node_index)
-                .map_err(|_| LibraryError::custom("Expected a valid node reference"))?;
-            leaf_references.push(node_ref);
+    /// Returns an iterator over a tuple of the leaf index and a reference to a
+    /// leaf, sorted according to their position in the tree from left to right.
+    pub(crate) fn leaves(&self) -> impl Iterator<Item = (LeafIndex, &T)> {
+        let mut original_leaves = self.original_tree.leaves().peekable();
+        let mut diff_leaves = self
+            .diff
+            .iter()
+            .filter_map(|(index, leaf)| {
+                if index % 2 == 0 {
+                    Some((*index / 2, leaf))
+                } else {
+                    None
+                }
+            })
+            .peekable();
+
+        let mut combined = Vec::new();
+
+        loop {
```

Comment:
I think getting rid of Itertools is nice, but `loop` is a bit dangerous. I seem to remember that we decided against using it in another instance due to the potential for an infinite loop, but I can't remember the specifics. This one seems pretty straight-forward due to how `next` is used in each iteration indicating that the loop must terminate at some point, so I don't have a strong opinion here.